### DISCUSSION TITLE
Update phase of current task in workflow controller "manually" to avoid workflow-controller/pod-deleted race condition

### DISCUSSION THREAD

**Discussion by clementemmanuel** [argoproj/argo-workflows#8957] [General]
I don't use argo directly but instead transitively via kubeflow.

When running under load a lot of tasks are (erroneously) marked in an error state `Error pod deleted`

I'm running a large amount workflows in parallel, these individual workflows are themselves large, these are all batch processes that are instigated and then run to completion.  
Rough numbers are something like thousands of pods running in parallel.

A lot of these pods are long-lived (~15 minutes), and I seem to be running up against a race condition between when the workflow controller attempts to update the state of a task/pod, and the GKE processes that delete the pod on completion.

I can see that the main container of our tasks run to completion succesfully, and I believe the argo wait container also completes.

When a pod completes succesfully, the pod is marked for deletion, occasionally however the workflow controller is unable to query the state of the associated pod fast enough, it will see that the pod was deleted and mark the phase as `error`, meaning we're unable to continue to any dependent tasks and the whole workflow is errored.

This only occurs when running under load, and I assume it's some combination of GKE needing to be more aggressive with deleting successful pods, and the workflow controller being slower needing to update more tasks.

The long-term solution would be for the workflow-controller to more appropriately manage the shutdown of a pod, claiming a finalizer that is appropriately applied and cleaned up, I believe this is being suggested here https://github.com/argoproj/argo-workflows/issues/8783

I also believe this is the same issue being seen in the following
* https://github.com/argoproj/argo-workflows/issues/7466
* https://github.com/argoproj/argo-workflows/issues/7994

I am trying to think of workarounds, but I can't seem to find any documentation going over how the workflow-controller pod in a cluster actually determines the phase of the tasks. I've tried manually patching running pods with `workflows.argoproj.io/completed: "true"`, but this just confuses the workflow controller which immediately updates the phase of the task from `running -> error` with `pod deleted`.

I am wondering what viable workarounds there would be, I'm trying to understand how I can eagerly report the phase of the task (i.e. from my main container, before exiting) to the workflow controller to avoid this race condition all-together.

Is something like this possible? Is there an alternative workaround that would work?

**Comment by tooptoop4** [comment]
@clementemmanuel this sounds like https://github.com/argoproj/argo-workflows/issues/10002#issuecomment-1379648441 I'm facing. did u solve?
